File: src/com/minhle/practiceleetcode/WordSearch.java
package com.minhle.practiceleetcode;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

public class WordSearch {

    static final int[] dR = {-1, 1, 0, 0};
    static final int[] dC = {0, 0, -1, 1};

    public boolean exist(char[][] board, String word) {
        int row = board.length;
        int column = board[0].length;

        List<int[]> possibleStartPoints = new ArrayList<>();
        for (int i = 0; i < row; i ++) {
            for (int j = 0; j < column; j++) {
                if (board[i][j] == word.charAt(0)) {
                    int[] point = new int[]{i,j};
                    possibleStartPoints.add(point);
                }
            }
        }

        for (int[] point : possibleStartPoints) {
            int[][] visited = new int[row][column];
            visited[point[0]][point[1]] = 1;
            if (dfs(board, visited, point, 1, word)) {
                return true;
            }
        }
        return false;
    }

    private static boolean dfs(char[][] map, int[][] visited, int[] startPoint, int pointer, String target) {
        if (pointer == target.length()) {
            return true;
        }

        boolean r = false;
        visited[startPoint[0]][startPoint[1]] = 1;
        for (int i = 0; i < 4; i++) {
            int nextX = startPoint[0] + dR[i];
            int nextY = startPoint[1] + dC[i];
            boolean inBound = nextX >= 0 && nextX < map.length && nextY >= 0 && nextY < map[0].length;
            if (inBound) {
                if (visited[nextX][nextY] == 0 && map[nextX][nextY] == target.charAt(pointer)) {
                    int[] nextStartPoint = new int[]{nextX, nextY};
                    r = r || dfs(map, visited, nextStartPoint, pointer + 1, target);
                }
            }
        }

        visited[startPoint[0]][startPoint[1]] = 0;
        return r;
    }

    @Test
    void firstTest() {
        char[][] map = {{'A','B','C','E'},
                        {'S','F','C','S'},
                        {'A','D','E','E'}};
        String target = "ABCCED";
        Assertions.assertTrue(exist(map, target));
    }

    @Test
    void secondTest() {
        char[][] map = {{'A','B','C','E'},
                        {'S','F','C','S'},
                        {'A','D','E','E'}};
        String target = "SEE";
        Assertions.assertTrue(exist(map, target));
    }

    @Test
    void lastTest() {
        char[][] map = {{'A','B','C','E'},
                        {'S','F','C','S'},
                        {'A','D','E','E'}};
        String target = "ABCB";
        Assertions.assertFalse(exist(map, target));
    }
}

File: src/com/minhle/practiceleetcode/WordSearch.cpp
#include "AllNecessaryHeaders.h"

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        string curr;
        for (int i = 0; i < board.size(); ++i)
        {
            for (int j = 0; j < board[i].size(); ++j)
            {
                if (board[i][j] == word.front())
                {
                    curr += word.front();
                    vector<vector<bool>> visited(board.size(), vector<bool>(board[0].size(), false));
                    visited[i][j] = true;
                    if (backtracking(board, word, curr, 1, i, j, visited))
                    {
                        return true;
                    }
                    curr.pop_back();
                }
            }
        }
        return false;
    }
private:
    const vector<int> dR = {0, 0, 1, -1};
    const vector<int> dC = {1, -1, 0, 0};

    bool backtracking(const vector<vector<char>> &board, const string &word, string &curr, int pos, int r, int c, vector<vector<bool>> &visited)
    {
        if (curr.size() == word.size())
        {
            return true;
        }
        if (r == board.size())
        {
            return false;
        }
        if (c == board[0].size())
        {
            return backtracking(board, word, curr, pos, r + 1, c, visited);
        }
        for (int i = 0; i < 4; ++i)
        {
            int next_r = r + dR[i];
            int next_c = c + dC[i];
            bool in_bound = next_r >= 0 && next_c >= 0 && next_r < board.size() && next_c < board[0].size();
            if (in_bound)
            {
                if (!visited[next_r][next_c] && board[next_r][next_c] == word[pos])
                {
                    curr += board[next_r][next_c];
                    visited[next_r][next_c] = true;
                    ++pos;
                    if (backtracking(board, word, curr, pos, next_r, next_c, visited))
                    {
                        return true;
                    }
                    --pos;
                    visited[next_r][next_c] = false;
                    curr.pop_back();
                }
            }
        }
        return false;
    }
};
